components/raftstore/src/store/fsm/peer.rs

```diff
@@ -1285,8 +1285,15 @@ where
                     self.maybe_destroy();
                 }
             }
-            CasualMessage::Campaign => {
-                let _ = self.fsm.peer.raft_group.campaign();
+            CasualMessage::Campaign { notify_by_parent } => {
+                if notify_by_parent {
+                    // If the message is sent by the parent, it means the parent is already the
+                    // leader. So this peer will be likely to become the leader
+                    // soon.
+                    let _ = self.fsm.peer.maybe_campaign(true);
```

Comment:
To avoid being triggered by arbitrarily slow parents, let's compare epoch and terms.

```rust
if parent.epoch == self.epoch && self.term == RAFT_INIT_LOG_TERM {
    let _ = self.fsm.peer.maybe_campaign(true);
}
```

Reply:
How about checking whether the current region has valid leader here?

Reply:
> How about checking whether the current region has valid leader here before campaign?

The campaign can still be triggered by slow-responding parents.

Reply:
If the new split region has valid leader, it won't trigger a new `campaign` here after introducing the check `if notify_by_parent && !self.fsm.peer.has_valid_leader()`